clang/lib/Sema/SemaExprCXX.cpp

```diff
@@ -9509,6 +9509,19 @@ Sema::ActOnStartRequiresExpr(SourceLocation RequiresKWLoc,
   PushDeclContext(BodyScope, Body);
 
   for (ParmVarDecl *Param : LocalParameters) {
+    if (Param->getType()->isVoidType()) {
+      if (LocalParameters.size() > 1) {
+        Diag(Param->getBeginLoc(), diag::err_void_only_param);
+        Param->setInvalidDecl();
+      } else if (Param->getIdentifier()) {
+        Diag(Param->getBeginLoc(), diag::err_param_with_void_type);
+        Param->setInvalidDecl();
+      } else if (Param->getType().hasQualifiers()) {
+        Diag(Param->getBeginLoc(), diag::err_void_param_qualified);
+        Param->setInvalidDecl();
```

Comment:
The error recovery which we apply for function parameters in these cases is to proceed with `int` as the parameter type, and not mark the parameters as invalid.
Why are we doing different here?

Reply:
Indeed performing a recovery would benefit source tools like clangd.

Probably, we can continue as if it were an int and give up evaluating the whole expression to false - again, this is an open question, and I have no answer :D

Reply:
The point of recovering to int would be to not give up, just assume the user intended to write int and proceed.